pennylane/templates/embeddings/iqp.py

```diff
@@ -0,0 +1,249 @@
+# Copyright 2018-2020 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+r"""
+Contains the ``IQPEmbedding`` template.
+"""
+# pylint: disable-msg=too-many-branches,too-many-arguments,protected-access
+from collections import Iterable
+from pennylane.templates.decorator import template
+from pennylane.ops import RZ, CNOT, Hadamard
+from pennylane.templates import broadcast
+from pennylane.templates.utils import (
+    check_shape,
+    check_wires,
+    check_type,
+    get_shape,
+    check_no_variable,
+)
+
+
+@template
+def zz(parameter, wires):
+    """Template for decomposition of ZZ coupling.
+
+    Args:
+        parameter (float): parameter of z rotation
+        wires (list[int]): qubit indices that the template acts on
+    """
+    CNOT(wires=wires)
+    RZ(2 * parameter, wires=wires[0])
+    CNOT(wires=wires)
+
+
+@template
+def IQPEmbedding(features, wires, n_repeats=1, pattern=None):
+    r"""
+    Encodes :math:`n` features into :math:`n` qubits using diagonal gates of an IQP circuit. The embedding
+    has been proposed by `Havlicek et al. (2018) <https://arxiv.org/pdf/1804.11326.pdf>`_. It and can be repeated.
+
+    .. warning::
+
+        ``IQPEmbedding`` calls a circuit that involves non-trivial classical processing of the
+        features. The ``features`` argument is therefore **not differentiable** when using the template, and
+        gradients with respect to the features cannot be computed by PennyLane.
```

Comment:
Yes, exactly. I thought about it, and it is a generic problem that we have everywhere. I think this needs an ADR...

Reply:
Agreed! It would be really cool to have this differentiable. I was wondering what the difference is between an IQPLayer and the IQPEmbedding, but I guess it's that we can't differentiate the features so it's better suited to an embedding.